numpy/core/include/numpy/ufuncobject.h

```diff
@@ -211,6 +211,13 @@ typedef struct _tagPyUFuncObject {
         /* Identity for reduction, when identity == PyUFunc_IdentityValue */
         PyObject *identity_value;
 
+        /* New private fields related to dispatching */
+        void *_dispatch_cache;
+        /*
+         * Currently, just a list, but that can never allow deletion (unless
+         * elements that may get deleted again are deleted occasionally).
+         */
```

Comment:
```suggestion
        /*
         * Currently, just a list, but that can never allow deletion (unless
         * elements that may get deleted again are deleted occasionally).
         */
```
```suggestion
        /* A PyListObject of (dtype, array-loop-function)  */
```

Reply:
Adopted as `/* A PyListObject of `(tuple of DTypes, ArrayMethod/Promoter)` */`

The TODO is a bit of a long story and very tricky.  So tending to skip it...

(The problem is what happens with registered or cached ufunc loops that become inaccessible if you *delete* a `DType` class.  But, I have yet to see a multiple-dispatching system dealing with that.  And I am not even sure that Python **can** deal with it; I think the GC is missing a feature necessary to do it properly anyway.)